README.md
# Proyecto_Scarleth
Curso Ingenieria de software, proyecto final

- Bryan Emanuel Paz Ramirez 
- Emmanuel Alexander Cabrera Alvarez
- Kendrick Rafael Vidal Ruano
- Julio Enrique Marroquin Tobar


CHANGELOG.md
*CHANGELONG PROYECTO SCARLETH*
*LOS HECHOS LOS ENCONTRARA EN ORDEN CRONOLOGICO*

*07/07/2023*
- Mejorando codigo de centras de actions modernizando el codigo utilizando funciones flecha.
- Agregando funcion buscar a los siguientes controladores y de la misma manera modernizadando un poco el codigo y agregando status code.
	- Cargos.
	- Direcciones.
	- Etnias.
	- Generos.
	- Medicos.
	- Pacientes.

*06/07/2023*
- mejorando el paginado mostrando el total de paginas, el record osea cuantos objetos contiene la pagina y la pagina actual.
- cambiando reply a helpers.
- agregando paginateReturn para ayudarnos a retornar la info del paginado como por ejemplo.
	- total de paginas.
	- total de objetos en pagina.
	- pagina actual.
	- data.

*05/07/2023*
- Mejorando Controladores agregando statuscodes.
- Agregando un helper de paginado.
- Agregando un context de la base de datos para no instanciarlo muchas veces.

*04/07/2023*
- Agregamos archivos como .env y .env-example, agregamos el .gitignore, mejoramos las variables de entorno.
- Agregamos seeds para poblar la base de datos de datos definidos como lo son genero o direccion.

*25/04/2023*
- Agregando controlador de Medico.
- Agregando actions de medico.
	- medico action.
	- Obtener medico.
	- Nuevo medico.

*19/04/2023*
- Agregamos una nueva configuracion y mejoramos la configuracion anterior de los actions y controladores.
- Iniciamos realizando pruebas.
- Agregamos por el momento los siguientes controladores con sus actions.
	- Pacientes.
	- Generos.
	- Etnia.
	- Direccio.
- Agregamos paginado.
	
*18/04/2023*
- Agregamos configuracion para iniciar con los action y los controladores.
- realizamos pruebas realizando consultas con postman.

*27/03/2023*
- Agregando data annotations a algunos modelos (terminar).
- Agregando todos los data annotations.
- Agregando las llaves fk y pk.
- Agregando entity para migraciones.
- Generamos y agregamos migraciones a la base de datos.

*25/03/2023*
- Terminamos de agregar los modelos.
- Agregamos el modelo ConexionContext para iniciar a configurar la conexion a la base de datos.
- Agregamos EntityFramework.
- Agregamos cadena de conexion.
- Configuramos el ConexionCOntext y program para tener la conexion con EntityFramework.
- Agregamos la nueva documentacion.

*22/03/2023*
- Terminamos los Skecth de muestra para como se veran los CRUD.

*19/03/2023*
- Realizamos mejora de la documentacion del sistema.
- Realizamos mejora de el ER.

*18/03/2023*
- Realizamos el ER de se tomara de referencia para la base de datos en donde manejaremos nuestros datos.
- Fase #1 dicha en el trello de documentacion finalizada.
- Mejorando el ER.
- Agreamos los modelos correspondientes del ER al api ya que estos se trabajaran con migraciones.

*16/03/2023*
- Realizamso el primer paquete para subirlo al github y empezar a cranear la base de datos.

*15/03/2023*
- Se Avanza y se culmina el UML 1.0 que es donde tenemos la idea de cada uno de los modulos que tendremos.

*08/03/2023*
- Se Avanza la documentacion.
- Se da seguimiento a los sketch.
- Craneamos una que se tendra para el UML.

*04/03/2023*
- Se Avanza en la documentacion.
- Se Inician los sketch.

*02/03/2023*
- Se Avanza en la documentacion.

*22/02/2023*
- Se realizo una encuesta para tener un poco de informacion sobre el lugar en donde brindaremos la ayuda.






app/Models/Medico.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using app.helpers;
namespace app.Models
{
    public class Medico : IControl_Fechas
    {
        [NotMapped]
        public const int ACTIVO = 1;
        [NotMapped]
        public const int NO_ACTIVO = 0;

        [Key]
        public int id { get; set; }
        
        [Required(ErrorMessage = $"{ErrorHelperMessage.campoRequired} [nombre]")]
        [StringLength(85, ErrorMessage = $"{ErrorHelperMessage.campoLenght} [85 caracteres] [nombre]")]
        [DataType(DataType.Text)]
        public string nombre { get; set; }
        
        [Required(ErrorMessage = $"{ErrorHelperMessage.campoRequired} [cui]")]
        [StringLength(15, ErrorMessage = $"{ErrorHelperMessage.campoLenght} [15 caracteres] [cui]")]
        [DataType(DataType.Text)]
        public string cui { get; set; }

        [Required(ErrorMessage = $"{ErrorHelperMessage.campoRequired} [numero colegiado]")]
        [StringLength(25, ErrorMessage = $"{ErrorHelperMessage.campoLenght} [25 caracteres] [numero colegiado]")]
        [DataType(DataType.Text)]
        public string numero_colegiado { get; set; }

        [Required(ErrorMessage = $"{ErrorHelperMessage.campoRequired} [fecha registro]")]
        [DataType(DataType.DateTime)]
        [DisplayFormat(DataFormatString = "{0:yyyy-MM-dd}")]
        public DateTime fecha_registro { get; set; }

        [Required(ErrorMessage = $"{ErrorHelperMessage.campoRequired} [telefono]")]
        [StringLength(20, ErrorMessage = $"{ErrorHelperMessage.campoLenght} [20 caracteres] [telefono]")]
        [DataType(DataType.PhoneNumber)]
        public string telefono { get; set; }

        [Required(ErrorMessage = $"{ErrorHelperMessage.campoRequired} [fecha nacimiento]")]
        [DataType(DataType.DateTime)]
        [DisplayFormat(DataFormatString = "{0:yyyy-MM-dd}")]
        public DateTime fecha_nacimiento { get; set; }
        
        [Required(ErrorMessage = $"{ErrorHelperMessage.campoRequired} [cargo]")]
        [ForeignKey("Cargo")]
        public int cargo_id { get; set; }

        [Required(ErrorMessage = $"{ErrorHelperMessage.campoRequired} [direccion]")]
        [ForeignKey("Direccion")]
        public int id_direccion { get; set; }

        [Required(ErrorMessage = $"{ErrorHelperMessage.campoRequired} [etnia]")]
        [ForeignKey("Etnia")]
        public int etnia_id { get; set; }

        [Required(ErrorMessage = $"{ErrorHelperMessage.campoRequired} [genero]")]
        [ForeignKey("Genero")]
        public int genero_id { get; set; }
        public int estado { get; set; } = ACTIVO;
        
        [Required(ErrorMessage = $"{ErrorHelperMessage.campoRequired} [usuario]")]
        [ForeignKey("Usuario")]
        public int usuario_id { get; set; }
        public DateTime CreateAt { get; set; } = DateTime.Now;
        public DateTime UpdateAt { get; set; }
        public DateTime DeleteAt { get; set; }

        [NotMapped]
        public virtual Cargo Cargo { get; set; }
        [NotMapped]
        public virtual Etnia Etnia { get; set; }
        [NotMapped]
        public virtual Genero Genero { get; set; }
        [NotMapped]
        public virtual Direccion Direccion { get; set; }
        [NotMapped]
        public virtual Usuario Usuario { get; set; }
    }
}